### ISSUE TITLE
Make Noddity smaller if at all possible

### ISSUE THREAD

**Issue by TehShrike** [TehShrike/noddity#27]
Disc: https://github.com/hughsk/disc

```sh
browserify --full-paths index.js | uglifyjs -mc | discify --open
```

1. check for including multiple versions of the same library
2. look for any low-hanging modules that could be easily dropped or replaced with something lighter

**Comment by ArtskydJ** [comment]
http://artskydj.github.io/test/noddity-disc.html

**Comment by TehShrike** [comment]
Got a bit cleaner version using `node node_modules/browserify/bin/cmd.js --full-paths /Users/josh/code/noddity/js/index.js | uglifyjs -mc | discify > ../noddity.html`: http://error420.com/noddity.html

**Comment by TehShrike** [comment]
First questions:

- why the hell is esprima required
- why the hell is autolinker required

**Comment by TehShrike** [comment]
Autolinker is used by Remarkable

**Comment by TehShrike** [comment]
esprima is used by js-yaml.  woooooow that hurts. @saibotsivad

**Comment by TehShrike** [comment]
As an aside, it looks that disc isn't quite cool at showing dependencies installed by the auto-flattening npm 3

**Comment by TehShrike** [comment]
Man, that 100k for front-matter support sucks

**Comment by TehShrike** [comment]
I'm going to leave this issue open for complaining/solution suggesting.

**Comment by ArtskydJ** [comment]
Maybe we can use [`https://www.npmjs.com/package/yaml-js`](https://www.npmjs.com/package/yaml-js) instead of `js-yaml`? Or we could use [`front-matter`](https://www.npmjs.com/package/front-matter) which uses `yaml-js`.

(In https://github.com/TehShrike/text-metadata-parser/blob/master/package.json)

**Comment by TehShrike** [comment]
man, yaml-js is even larger than js-yaml

**Comment by ArtskydJ** [comment]
The `yaml` package looks promising. [GitHub](https://github.com/tj/js-yaml)/[npm](https://www.npmjs.com/package/yaml). It is 4.2 kb minified. However, the maintenance on that package looks minimal.

**Comment by saibotsivad** [comment]
That `yaml` package has zero tests and is really old, so that would be kind of scary to use, I think.

**Comment by ArtskydJ** [comment]
According to [this](https://github.com/nodeca/js-yaml/blob/master/CHANGELOG.md#300--2013-12-16) `js-yaml` shouldn't be including esprima in the browser version.

See https://github.com/nodeca/js-yaml/pull/231.

**Comment by ArtskydJ** [comment]
I don't know if this package is worth looking into...
https://github.com/jonschlinkert/gray-matter
(uses js-yaml under the hood)

**Comment by TehShrike** [comment]
I think once this js-yaml fix is deployed https://github.com/nodeca/js-yaml/issues/230 it will be fine as a yaml parser.  30KiB is still non-trivial, but I don't know of a better solution without falling back to the original text-metadata-parser implementation, or writing a bunch of tests for tj's old yaml project and using it to support a subset of yaml.

**Comment by ArtskydJ** [comment]
js-yaml@3.5.1 has the fix.

**Comment by TehShrike** [comment]
Now 60KiB lighter! [old](http://error420.com/noddity-2015-11-24.html) vs [new](http://error420.com/noddity-2016-01-20.html)

**Comment by saibotsivad** [comment]
Hot diggity!

On January 20, 2016 8:44:32 AM CST, Josh Duff <notifications@github.com> wrote:
>Now 60KiB lighter! [old](http://error420.com/noddity-2015-11-24.html)
>vs [new](http://error420.com/noddity-2016-01-20.html)
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/TehShrike/noddity/issues/27#issuecomment-173224722

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by ArtskydJ** [comment]
How are these getting included?

- readable-stream
- buffer

**Comment by ArtskydJ** [comment]
With npm@2 it was easy to see what was including what. npm@3 makes this quite difficult. :/

**Comment by TehShrike** [comment]
Yeah, and it doesn't have a "flatten like how npm@2 would" option that I can find either :-\

**Comment by ArtskydJ** [comment]
The easiest way that I can think of is to use nvm to switch to node 0.10/npm 2.x, (if it switches both at once,) and then install the deps.

Probably not worth redoing the current files, but maybe something to try in the future?

**Comment by sundarj** [comment]
Use `npm ls`:

```                                                                                                                                  
/m/t/d/k/m/noddity (master) λ npm ls
noddity@4.3.0 /media/tumnus/doc/kaihatsu/master/noddity
├─┬ browserify@10.2.6
│ ├── assert@1.3.0
│ ├─┬ browser-pack@5.0.1
│ │ ├─┬ combine-source-map@0.6.1
│ │ │ ├── convert-source-map@1.1.3
│ │ │ ├── inline-source-map@0.5.0
│ │ │ ├── lodash.memoize@3.0.4
│ │ │ └── source-map@0.4.4
│ │ ├── defined@1.0.0
│ │ └── umd@3.0.1
│ ├── browser-resolve@1.11.1
│ ├─┬ browserify-zlib@0.1.4
│ │ └── pako@0.2.8
│ ├─┬ buffer@3.6.0
│ │ ├── base64-js@0.0.8
│ │ ├── ieee754@1.1.6
│ │ └── isarray@1.0.0
│ ├── builtins@0.0.7
│ ├── commondir@0.0.1
│ ├─┬ concat-stream@1.4.10
│ │ └── typedarray@0.0.6
│ ├─┬ console-browserify@1.1.0
│ │ └── date-now@0.1.4
│ ├── constants-browserify@0.0.1
│ ├─┬ crypto-browserify@3.11.0
│ │ ├─┬ browserify-cipher@1.0.0
│ │ │ ├─┬ browserify-aes@1.0.6
│ │ │ │ └── buffer-xor@1.0.3
│ │ │ ├─┬ browserify-des@1.0.0
│ │ │ │ └─┬ des.js@1.0.0
│ │ │ │   └── minimalistic-assert@1.0.0
│ │ │ └── evp_bytestokey@1.0.0
│ │ ├─┬ browserify-sign@4.0.0
│ │ │ ├── bn.js@4.11.3
│ │ │ ├── browserify-rsa@4.0.1
│ │ │ ├─┬ elliptic@6.2.3
│ │ │ │ ├── brorand@1.0.5
│ │ │ │ └── hash.js@1.0.3
│ │ │ └─┬ parse-asn1@5.0.0
│ │ │   └── asn1.js@4.5.2
│ │ ├── create-ecdh@4.0.0
*snip*
```

**Comment by TehShrike** [comment]
Nice.  Now we just need a way to correlate each dependency to how much they're contributing to the browserified output

**Comment by sundarj** [comment]
Would comparing their folder-sizes not do that?

**Comment by TehShrike** [comment]
It would not.  A lot of those dependencies are only there for tests/building.  You can't even browserify the `main` entry in the package.json, since some modules may be depending on specific files inside of the modules and not even requiring the module name by itself.

**Comment by sundarj** [comment]
If you do `npm install --production` that should get rid of most of the tests/building packages (at least in theory) - but yeah, the second point is also true.

**Comment by TehShrike** [comment]
I don't know of any other easy wins, so I'm going to close this for now.

**Comment by TehShrike** [comment]
Apparently noddity has bloated to nearly 600KiB.  http://error420.com/noddity-2016-06-21.html

**Comment by TehShrike** [comment]
In case anyone is curious, this is how I got an unflattened node_modules installation without installing npm@2 globally:

```sh
rm -rf node_modules
npm i npm@2
node node_modules/.bin/npm i
```

**Comment by TehShrike** [comment]
After publishing noddity-retrieval@1.5.3: http://error420.com/noddity-2016-06-24.html

The next easiest dependency to drop is probably js-yaml in text-metadata-parser: https://github.com/TehShrike/text-metadata-parser/issues/12

Finding out where Buffer is coming from would save us 20KiB.

Dropping [ent](https://www.npmjs.com/package/ent) from noddity-render-dom might be nice too if we could find a way.

Remarkable is still pretty large, but I don't know if we can move away from it at this point.

**Comment by ArtskydJ** [comment]
`ent` is large because of entities.json, which is used for decoding named entities. Without entities.json, it would be unable to decode `&nbsp;` but would be able to decode `&#160;`.

We're using `ent` [here](https://github.com/TehShrike/noddity-template-parser/blob/master/htmlify.js#L26-L27) to undo stuff that remarkable encoded. Maybe we don't need the named entity decoding? If we don't, are we able to use browserify to replace entities.json with `{}`?

**Comment by ArtskydJ** [comment]
For reference:

- noddity-2016-06-24 (487kb)
- ractive (181kb)
- remarkable (88kb)
- js-yaml (38kb)
- ent (32kb)
- levelup (27kb)
- buffer (20kb)
- readable-stream (16kb)
- fruitdown (16kb)
- superagent (10kb)
- subleveldown (10kb)
- **total** (440kb)

Everything else looked like it was under 10kb.
